packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ContainerRuntimeMessage } from "..";
+import { IBatch } from "./definitions";
+
+interface IGroupedMessage {
+	contents: string;
+	metadata: Record<string, unknown> | undefined;
+	compression: string;
+}
+
+export class OpGroupingManager {
+	static groupedBatchOp = "groupedBatch";
+
+	constructor(private readonly groupedBatchingEnabled: boolean) {}
+
+	public groupBatch(batch: IBatch): IBatch {
+		if (batch.content.length < 2 || !this.groupedBatchingEnabled) {
+			return batch;
+		}
+
+		// Need deserializedContent for back-compat
+		const deserializedContent = {
+			type: OpGroupingManager.groupedBatchOp,
+			contents: batch.content.map((message) => ({
+				contents: message.contents,
+				metadata: message.metadata,
+				compression: message.compression,
+			})),
+		};
+
+		const groupedBatch: IBatch = {
+			...batch,
+			content: [
+				{
+					...batch.content[0],
```

Comment:
This looks very suspicious. Why would you use ... here (as well as couple lines above)?
I believe you only getting referenceSequenceNumber & localOpMetadata. And I think latter can be removed as this this format is used only for sending ops to Loader layer, it's not used in op resending / rebasing logic, thus localOpMetadata is not useful.
referenceSequenceNumber is also weird (thought it's not related to your changes) - I'd like @andre4i to make a call here - are we enforcing same reference sequence number for a batch, or not? As currently we have reference sequence on both IBatch & IBatchMessage, which does not make much sense.